### PR TITLE
BSIP 75 - Asset Owner defines MCR and MSSR values

### PR DESCRIPTION
**Pull request by jmjatlanta**
See Issue #96

- [ ] Add specifications
- [ ] Verify rationale
- [ ] Research risks

### CODE REVIEW

File: bsip-0075.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+# Abstract
+This BSIP provides the ability for the asset creator to decide if adjustments to the Maintenance Collateral Ratio (MCR) or the Maximum Short Squeeze Ratio (MSSR) should be done by the asset owner or by price feed providers.
+
+# Motivation
+Asset owners are often in the best position to control the values of MCR and MSSR of their asset. Therefore, they should be able to have direct control to adjust those values, should they choose to allow themselves.
+
+# Rationale
+Prior to this change, adjustments to MCR and MSSR were handled by feed producers. This often requires the asset owner to contact the feed producer and ask for the change. For some assets, such a step is time consuming for both parties and ill-fitted as part of the feed producer's responsibilities.
+
+This new change shall permit the asset owner to determine at asset creation time which should have the responsibility to maintain those values.
```
Comment (MichelSantos) [inline bsip-0075.md]: I might emphasize "at creation time", and change "which" to "who"

File: bsip-0075.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+# Specification
+The `asset_object.asset_options.flags` shall contain the new flag `issuer_fed_mcr` that determines if the MCR value is stored as part of the asset_object, or if it is pulled from the feeds.
+
+The `asset_object.asset_options.flags` shall contain the new flag `issuer_fed_mssr` that determines if the MSSR value is stored as part of the asset_object, or if it is pulled from the feeds.
+
+The `create_asset_operation` shall be modified to include the setting of the parameters above, as well as initial values for MCR and MSSR.
+
+The `update_asset_operation` shall be modified permit updating of the MCR and MSSR values, if the asset has been created with the permission for the asset owner to do so. NOTE: much of the logic currently in `asset_publish_feeds_evaluator::do_apply()` (i.e. call order checks and calculations) shall be also called here if MCR or MSSR are adjusted.
```
Comment (MichelSantos) [inline bsip-0075.md]: shall be modified **to** permit

File: bsip-0075.md [hunk 3]
```diff
@@ -20,16 +20,16 @@ Asset owners are often in the best position to control the values of MCR and MSS
...
+The `asset_object.asset_options.issuer_permissions` shall contain the new `issuer_fed_mssr` permission that determines if an issuer can change the related `flags` value.
 
-The `create_asset_operation` shall be modified to include the setting of the parameters above, as well as initial values for MCR and MSSR.
+The `create_asset_operation` shall be modified to include the setting of the parameters above, as well as initial `flags` values for MCR and MSSR.
 
-The `update_asset_operation` shall be modified permit updating of the MCR and MSSR values, if the asset has been created with the permission for the asset owner to do so. NOTE: much of the logic currently in `asset_publish_feeds_evaluator::do_apply()` (i.e. call order checks and calculations) shall be also called here if MCR or MSSR are adjusted.
+The `update_asset_operation` shall be modified to permit updating of the MCR and MSSR values, if the asset has the appropriate `flags` value set permitting the asset owner to do so. NOTE: much of the logic currently in `asset_publish_feeds_evaluator::do_apply()` (i.e. call order checks and calculations) shall be also called here if MCR or MSSR are adjusted.
```
Comment (MichelSantos) [inline bsip-0075.md]: I think that the prior text, which references the pre-existing permission setting of the asset, is more correct

File: bsip-0075.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+# Motivation
+Asset owners are often in the best position to control the values of MCR and MSSR of their asset. Therefore, they should be able to have direct control to adjust those values, should they choose to allow themselves.
+
+# Rationale
+Prior to this change, adjustments to MCR and MSSR were handled by feed producers. This often requires the asset owner to contact the feed producer and ask for the change. For some assets, such a step is time consuming for both parties and ill-fitted as part of the feed producer's responsibilities.
+
+This new change shall permit the asset owner to determine *at asset creation time* who should have the responsibility to maintain those values.
+
+# Specification
```
Comment (abitmore) [inline bsip-0075.md]: I think we need to extend an operation so the asset owner is able to specify the MCR and MSSR she desired.

Comment (MichelSantos) [reply bsip-0075.md]: @abitmore Do you recommend adding that detail to the text below that discusses those operations?

Comment (jmjatlanta) [reply bsip-0075.md]: I think that is a good idea too. I will add it after reviewing the code implications.

Comment (jmjatlanta) [reply bsip-0075.md]: Added. Please take a look.

File: bsip-0075.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+# Abstract
+This BSIP provides the ability for the asset creator to decide if adjustments to the Maintenance Collateral Ratio (MCR) or the Maximum Short Squeeze Ratio (MSSR) should be done by the asset owner or by price feed providers.
+
+# Motivation
+Asset owners are often in the best position to control the values of MCR and MSSR of their asset. Therefore, they should be able to have direct control to adjust those values, should they choose to allow themselves.
+
+# Rationale
+Prior to this change, adjustments to MCR and MSSR were handled by feed producers. This often requires the asset owner to contact the feed producer and ask for the change. For some assets, such a step is time consuming for both parties and ill-fitted as part of the feed producer's responsibilities.
+
+This new change shall permit the asset owner to determine *at asset creation time* who should have the responsibility to maintain those values.
```
Comment (grctest) [inline bsip-0075.md]: How will existing smartcoins be affected?

Comment (abitmore) [reply bsip-0075.md]: Existing smartcoins aren't affected. The owners can change settings after the protocol upgrade though.

**Comment by abitmore** [comment]
I wonder why this hasn't been merged.

**Comment by abitmore** [comment]
I got a better idea about specification and implementation lately: just add new optional value fields in bitasset options, which can be updated by the asset owners:
```
struct bitasset_options 
{
   //...
   optional<uint16_t> mcr;
   optional<uint16_t> mssr;
   optional<uint16_t> icr;
};
```
If one of the parameter is set, it means the owner has an opinion on the parameter and it will be in effect; if not set, then the value provided by the feed producers will be in effect.

We can assign a permission bit on each parameter too, so the owners can decide whether to forfeit the right to update it.

**Comment by MichelSantos** [comment]
> We can assign a permission bit on each parameter too, so the owners can decide whether to forfeit the right to update it.

I proposed a commit to include permissions.  I left out references to ICR because I thought that it might be out of scope.  If appropriate I have text which to update BSIP77 with a similar discussion about permissions.

**Comment by abitmore** [comment]
Yes I think BSIP77 can be updated similarly. Thanks.

**Pull request merged**